src/Polysemy/Writer.hs

```diff
@@ -122,3 +127,122 @@ runWriterAssocR =
     {-# INLINE go #-}
   in fmap (first ($ mempty)) . go
 {-# INLINE runWriterAssocR #-}
+
+------------------------------------------------------------------------------
+-- | Run a 'Writer' effect by transforming it into atomic
+-- operations over an 'IORef'.
+--
+-- /Note/: All 'tell's in the argument of a 'listen' or 'pass' will temporarily
+-- be stored seperately from the provided 'IORef' until the 'listen'\/'pass'
+-- completes.
+-- Therefore, such 'tell's will be lost if the argument to the 'listen'\/'pass'
+-- fails due to any effect interpreted after 'runWriterIORef', or due to an
+-- 'IO' exception.
+--
+-- In addition, any 'tell's of any asynchronous computation spawned inside a
+-- 'listen'\/'pass' will be lost once the 'listen'\/'pass' completes.
+-- 'runWriterTVar' avoids this particular issue.
+runWriterIORef
+  :: (Monoid o, Member (Embed IO) r)
+  => IORef o
+  -> Sem (Writer o ': r) a
+  -> Sem r a
+runWriterIORef ref = interpretH $ \case
+  Tell o -> do
+    t <- embed $ atomicModifyIORef' ref (\s -> (s <> o, ()))
+    pureT t
+  Listen m -> do
+    mm   <- runT m
+    -- TODO(KingoftheHomeless): this is REALLY stupid
+    ref' <- embed $ newIORef mempty
+    fa   <- raise $ runWriterIORef ref' mm
+    o    <- embed $ readIORef ref'
+    embed $ atomicModifyIORef' ref (\s -> (s <> o, ()))
+    return $ fmap (o, ) fa
+  Pass m -> do
+    mm   <- runT m
+    ins  <- getInspectorT
+    ref' <- embed $ newIORef mempty
+    t    <- raise $ runWriterIORef ref' mm
+    let f = maybe id fst $ inspect ins t
+    o    <- embed $ readIORef ref'
+    embed $ atomicModifyIORef' ref (\s -> (s <> f o, ()))
+    return (fmap snd t)
+
+------------------------------------------------------------------------------
+-- | Run a 'Writer' effect by transforming it into atomic operations over a
+-- 'TVar'.
+--
+-- /Note/: All 'tell's in the argument of a 'listen' or 'pass' will temporarily
+-- be stored seperately from the provided 'TVar' until the 'listen'\/'pass'
+-- completes.
+-- Therefore, such 'tell's will be lost if the argument to the 'listen'\/'pass'
+-- fails due to any effect interpreted after 'runWriterIORef',
+-- or due to an 'IO' exception.
+runWriterTVar :: (Monoid o, Member (Embed IO) r)
+              => TVar o
+              -> Sem (Writer o ': r) a
+              -> Sem r a
+runWriterTVar tvar = runWriterTVarAction $ \o -> do
+  s <- readTVar tvar
+  writeTVar tvar $! s <> o
+{-# INLINE runWriterTVar #-}
+
+runWriterTVarAction :: forall o a r
+                     . (Monoid o, Member (Embed IO) r)
+                    => (o -> STM ())
+                    -> Sem (Writer o ': r) a
+                    -> Sem r a
+runWriterTVarAction act = interpretH $ \case
```

Comment:
I'm interested to hear what you think about this. The implementation is a mess, but it gives us pretty good semantics.